File: validation/roundtrip/probe.py
#!/usr/bin/env python3
"""Behavioral prober — Python twin of probe.js (for vantage hosts without
Node). Same scenarios and report format, so compare.js works across reports
from either prober.

  python3 probe.py --vip <ip[:port]> [--https] [--label X] [--requests N]
                   [--admin id=host:port,...] [--hc-wait S] [--out report.json]
"""
import sys, json, time, ssl, urllib.request, urllib.error

opt = {'requests': 100, 'persist_reqs': 20, 'hc_wait': 20, 'label': 'lb',
       'out': None, 'https': False, 'admin': {}, 'vip': None}
args = sys.argv[1:]
i = 0
while i < len(args):
    a = args[i]
    if a == '--vip': i += 1; opt['vip'] = args[i]
    elif a == '--https': opt['https'] = True
    elif a == '--label': i += 1; opt['label'] = args[i]
    elif a == '--requests': i += 1; opt['requests'] = int(args[i])
    elif a == '--hc-wait': i += 1; opt['hc_wait'] = int(args[i])
    elif a == '--out': i += 1; opt['out'] = args[i]
    elif a == '--admin':
        i += 1
        for kv in args[i].split(','):
            k, v = kv.split('='); opt['admin'][k] = v
    i += 1
if not opt['vip']:
    print('usage: probe.py --vip <ip[:port]> [options]'); sys.exit(2)

host, _, port = opt['vip'].partition(':')
port = port or ('443' if opt['https'] else '80')
base = ('https' if opt['https'] else 'http') + '://' + host + ':' + port
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

def get(path, cookie=None):
    req = urllib.request.Request(base + path)
    if cookie:
        req.add_header('Cookie', cookie)
    try:
        with urllib.request.urlopen(req, timeout=5, context=ctx if opt['https'] else None) as r:
            return {'status': r.status, 'backend': r.headers.get('X-Backend'),
                    'set_cookie': r.headers.get_all('Set-Cookie') or []}
    except urllib.error.HTTPError as e:
        return {'status': e.code, 'backend': e.headers.get('X-Backend'), 'set_cookie': []}
    except Exception as e:
        return {'error': type(e).__name__}

def admin(hostport, path):
    h, _, p = hostport.partition(':')
    try:
        req = urllib.request.Request('http://%s:%s%s' % (h, p or '8080', path), method='POST', data=b'')
        with urllib.request.urlopen(req, timeout=5) as r:
            return r.status
    except Exception as e:
        return 'ERR ' + type(e).__name__

def distribution(n):
    counts, errors = {}, 0
    for k in range(n):
        r = get('/?i=%d' % k)
        if r.get('error') or not r.get('backend'):
            errors += 1
        else:
            counts[r['backend']] = counts.get(r['backend'], 0) + 1
    return {'requests': n, 'errors': errors, 'counts': counts}

def persistence():
    first = get('/persist-start')
    if first.get('error') or not first.get('backend'):
        return {'skipped': 'no backend header on first request', 'error': first.get('error')}
    cookie = '; '.join(c.split(';')[0] for c in first['set_cookie'])
    if not cookie:
        return {'skipped': 'LB did not set a cookie (cookie persistence not configured on this service?)'}
    same = other = errors = 0
    for k in range(opt['persist_reqs']):
        r = get('/persist?i=%d' % k, cookie)
        if r.get('error') or not r.get('backend'):
            errors += 1
        elif r['backend'] == first['backend']:
            same += 1
        else:
            other += 1
    return {'firstBackend': first['backend'], 'cookie': cookie.split('=')[0],
            'stuck': same, 'moved': other, 'errors': errors,
            'pass': other == 0 and same > 0}

def health():
    ids = list(opt['admin'].keys())
    if not ids:
        return {'skipped': 'no --admin endpoints given'}
    victim = ids[0]
    before = distribution(30)
    admin(opt['admin'][victim], '/admin/down')
    time.sleep(opt['hc_wait'])
    during = distribution(30)
    admin(opt['admin'][victim], '/admin/up')
    time.sleep(opt['hc_wait'])
    after = distribution(30)
    return {'victim': victim, 'before': before['counts'], 'during': during['counts'],
            'after': after['counts'], 'duringErrors': during['errors'],
            'ejected': victim not in during['counts'],
            'restored': victim in after['counts'],
            'pass': victim not in during['counts'] and victim in after['counts'] and during['errors'] == 0}

report = {'label': opt['label'], 'vip': opt['vip'], 'https': opt['https'],
          'when': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())}
print('[%s] distribution (%d requests)...' % (opt['label'], opt['requests']))
report['distribution'] = distribution(opt['requests'])
print('  ', json.dumps(report['distribution']['counts']), 'errors=%d' % report['distribution']['errors'])
print('[%s] persistence...' % opt['label'])
report['persistence'] = persistence()
print('  ', json.dumps(report['persistence']))
print('[%s] health-check reaction...' % opt['label'])
report['health'] = health()
print('  ', json.dumps({k: report['health'].get(k) for k in ('ejected', 'restored', 'pass', 'skipped')}))
if opt['out']:
    with open(opt['out'], 'w') as f:
        json.dump(report, f, indent=2)
    print('wrote ' + opt['out'])

File: validation/roundtrip/probe.js
#!/usr/bin/env node
'use strict';
// Behavioral prober for the F5-vs-Alteon round-trip comparison.
// Run the SAME probe against the app behind the F5 and behind the migrated
// Alteon, then diff the two reports with compare.js.
//
//   node probe.js --vip <ip[:port]> [--https] [--label f5|alteon]
//                 [--requests 100] [--backends id1,id2,...]
//                 [--admin id=host:port,...]   (direct backend admin endpoints)
//                 [--hc-wait 20]               (seconds to wait for LB to react)
//                 [--out report.json]
//
// Scenarios (each skipped gracefully if its prerequisites are missing):
//   distribution — N plain requests; tally X-Backend per response
//   persistence  — carry cookies from the first response; M follow-ups must
//                  hit the same backend (cookie persistence)
//   health       — mark one backend down via its DIRECT admin endpoint, wait,
//                  verify the LB stops sending traffic to it; bring it back,
//                  verify it returns to rotation
//   tls          — HTTPS handshake succeeds and traffic completes (--https)
const fs = require('fs');
const http = require('http');
const https = require('https');

const args = process.argv.slice(2);
const opt = { requests: 100, persistReqs: 20, hcWait: 20, label: 'lb', out: null, httpsMode: false, admin: {} };
for (let i = 0; i < args.length; i++) {
  const a = args[i];
  if (a === '--vip') opt.vip = args[++i];
  else if (a === '--https') opt.httpsMode = true;
  else if (a === '--label') opt.label = args[++i];
  else if (a === '--requests') opt.requests = parseInt(args[++i], 10);
  else if (a === '--backends') opt.backends = args[++i].split(',');
  else if (a === '--hc-wait') opt.hcWait = parseInt(args[++i], 10);
  else if (a === '--out') opt.out = args[++i];
  else if (a === '--admin') for (const kv of args[++i].split(',')) { const [id, hp] = kv.split('='); opt.admin[id] = hp; }
}
if (!opt.vip) { console.error('usage: node probe.js --vip <ip[:port]> [options]'); process.exit(2); }
const [vipHost, vipPort] = opt.vip.includes(':') ? opt.vip.split(':') : [opt.vip, opt.httpsMode ? '443' : '80'];

function get(pathname, extra) {
  const mod = opt.httpsMode ? https : http;
  return new Promise((resolve) => {
    const req = mod.request({
      host: vipHost, port: +vipPort, path: pathname, method: 'GET',
      headers: (extra && extra.headers) || {}, rejectUnauthorized: false, timeout: 5000
    }, (res) => {
      let body = '';
      res.on('data', (c) => body += c);
      res.on('end', () => resolve({ status: res.statusCode, backend: res.headers['x-backend'] || null, setCookie: res.headers['set-cookie'] || [], body }));
    });
    req.on('timeout', () => { req.destroy(); resolve({ error: 'timeout' }); });
    req.on('error', (e) => resolve({ error: e.code || e.message }));
    req.end();
  });
}

function adminPost(hostport, pathname) {
  const [h, p] = hostport.split(':');
  return new Promise((resolve) => {
    const req = http.request({ host: h, port: +(p || 8080), path: pathname, method: 'POST', timeout: 5000 },
      (res) => { res.resume(); res.on('end', () => resolve(res.statusCode)); });
    req.on('error', (e) => resolve('ERR ' + (e.code || e.message)));
    req.on('timeout', () => { req.destroy(); resolve('ERR timeout'); });
    req.end();
  });
}

const sleep = (s) => new Promise((r) => setTimeout(r, s * 1000));

async function distribution(n) {
  const counts = {}; let errors = 0;
  for (let i = 0; i < n; i++) {
    const r = await get('/?i=' + i);
    if (r.error || !r.backend) errors++;
    else counts[r.backend] = (counts[r.backend] || 0) + 1;
  }
  return { requests: n, errors, counts };
}

async function persistence() {
  const first = await get('/persist-start');
  if (first.error || !first.backend) return { skipped: 'no backend header on first request', error: first.error };
  const cookie = first.setCookie.map((c) => c.split(';')[0]).join('; ');
  if (!cookie) return { skipped: 'LB did not set a cookie (cookie persistence not configured on this service?)' };
  let same = 0, other = 0, errors = 0;
  for (let i = 0; i < opt.persistReqs; i++) {
    const r = await get('/persist?i=' + i, { headers: { Cookie: cookie } });
    if (r.error || !r.backend) errors++;
    else if (r.backend === first.backend) same++;
    else other++;
  }
  return { firstBackend: first.backend, cookie: cookie.split('=')[0], stuck: same, moved: other, errors, pass: other === 0 && same > 0 };
}

async function health() {
  const ids = Object.keys(opt.admin);
  if (!ids.length) return { skipped: 'no --admin endpoints given' };
  const victim = ids[0];
  const before = await distribution(30);
  await adminPost(opt.admin[victim], '/admin/down');
  await sleep(opt.hcWait);
  const during = await distribution(30);
  await adminPost(opt.admin[victim], '/admin/up');
  await sleep(opt.hcWait);
  const after = await distribution(30);
  return {
    victim,
    before: before.counts, during: during.counts, after: after.counts,
    duringErrors: during.errors,
    ejected: !(victim in during.counts),
    restored: victim in after.counts,
    pass: !(victim in during.counts) && (victim in after.counts) && during.errors === 0
  };
}

(async () => {
  const report = { label: opt.label, vip: opt.vip, https: opt.httpsMode, when: new Date().toISOString() };
  console.log('[' + opt.label + '] distribution (' + opt.requests + ' requests)...');
  report.distribution = await distribution(opt.requests);
  console.log('  ', JSON.stringify(report.distribution.counts), 'errors=' + report.distribution.errors);
  console.log('[' + opt.label + '] persistence...');
  report.persistence = await persistence();
  console.log('  ', JSON.stringify(report.persistence));
  console.log('[' + opt.label + '] health-check reaction...');
  report.health = await health();
  console.log('  ', JSON.stringify({ ejected: report.health.ejected, restored: report.health.restored, pass: report.health.pass, skipped: report.health.skipped }));
  if (opt.out) { fs.writeFileSync(opt.out, JSON.stringify(report, null, 2)); console.log('wrote ' + opt.out); }
})();
